### PR TITLE
Fix #4411 Improves building from source documentation

### LINKED ISSUE #4411: Improve "building from source" documentation for Linux
I've had at least a few people trying to install Astropy on Linux not realizing that a missing "Python.h" means they need to install the python development package for their system.  The current documentation simply reads 

> You will need a compiler suite and the development headers for Python and Numpy in order to build Astropy. On Linux, using the package manager for your distribution will usually be the easiest route

and then goes on to give lots of OSX-specific instructions.  It does not give any more information for Linux.  In the past I've been hesitant to pile on too much information for Linux because 1) An implicit assumption (often incorrect) that Linux users can be expected to "know what they're dong" and 2) It's hard to give comprehensive instructions for different Linux distros.

However, the existing text is horribly vague even under that rubric.  It should at least mention that most distros have a separate Python development package called either "python-dev" or "python-devel".  It also couldn't hurt to supply examples with apt-get and yum at a minimum.

While we're at it I think that section should also be reorganized.  It should open with some general text, and then be broken into subsections specific to OSX, Linux, and ~~Windows~~.  This will make the section less cluttered, and easier to navigate to the information a specific user needs for their platform.  This will at least address a small part of #1807.

**Update:** As I noted in #4420 there are not any existing build instructions specific to Windows, so implementing this task should not include that, as it's a non-trivial undertaking.  Instead this task should include just:
* [ ] Reorganize the existing text so that OS-specific instructions are corralled into OS-specific subsections, and
* [ ] For the Linux specific section add the aforementioned requirement of installing the OS's Python-dev package.

### PR DESCRIPTION
**Pull request by sudheesh001**

### CODE REVIEW

**Comment by embray** [comment]
Thanks. This is a start but doesn't fully accomplish what I wanted in #4411.  I also wanted that entire section "Building from source" rewritten so that the intro to the section does not give any OS-specific instructions, and intead contains subsections (at least for OSX, Linux, and Windows) where OS-specific instructions are contained.  The intro should tell users to see instructions specific to their platform.  As it is the text mixes up different instructions for Linux and OSX in arbitrary ways, making it very confusing to read.

For now actually leave Windows out since there are not currently any Windows-specific instructions in that section, and I will need to write them.

**Comment by sudheesh001** [comment]
@embray Does this solve the #4411 checkboxes ?

**Comment by embray** [comment]
Thanks @sudheesh001 that looks great.

My one question is whether we should be suggesting users install Homebrew in order to build on OSX.  Is that normal setup for Python development on OSX?  I want to caution against users installing things like Homebrew if it can be avoided, but I don't know enough to know what the best advice is here.  I'd like to ask about this on the mailinglist before proceeding.

Other than that though this is exactly what I had in mind, thanks.

File: docs/install.rst [hunk 1]
```diff
@@ -168,6 +162,45 @@ to build from source, unless you are installing a numbered release. (The
...
+For installing XCode command line tools::
+    
+    xcode-select --install
+
+and follow the onscreen instructions to install the command line tools required.
+
+You'll also need the python development package for Mac OS X to proceed in building
+astropy from source.
+
+To install python development package for Mac OS X,
```
Comment (embray) [inline docs/install.rst]: "To install python" -> "To install the python"

File: docs/install.rst [hunk 2]
```diff
@@ -168,6 +162,45 @@ to build from source, unless you are installing a numbered release. (The
 releases packages have the necessary C files packaged with them, and hence do
 not require Cython.)
 
+Prerequisites for Linux
+-----------------------
+
+On Linux, using the package manager for your distribution will usually be 
+the easiest route. In order to build from source, you'll need the python development package
+for your distro. You can install before proceeding with the setup process using
```
Comment (embray) [inline docs/install.rst]: "You can install before proceeding with the setup process using"  <- Strike this sentence entirely.  I think it's superfluous.

**Comment by astrofrog** [comment]
> My one question is whether we should be suggesting users install Homebrew in order to build on OSX. Is that normal setup for Python development on OSX? I want to caution against users installing things like Homebrew if it can be avoided, but I don't know enough to know what the best advice is here. I'd like to ask about this on the mailinglist before proceeding.

I don't think we should recommend installing Homebrew by default. If we want to mention it, we should also mention MacPorts too for example.

**Comment by sudheesh001** [comment]
> I don't think we should recommend installing Homebrew by default. If we want to mention it, we should also mention MacPorts too for example.

I couldn't figure out how to install using macports so i included only homebrew, what do you think is the best way  for macports.

**Comment by bsipocz** [comment]
e.g. ``fink install python35``, not sure about the macports equivalent but it should be something along the line of ``port install python35``. Or I guess it can be installed directly from https://www.python.org/downloads/mac-osx/ or using conda/miniconda.

Anyway going into this much detail may be superfluous anyway, the important point is that a newer python that the OSX stock one needs to be installed, and one usually knows their preferred package manager.

**Comment by mwcraig** [comment]
I think the most important thing to mention for OSX is that people shouldn't use system python. They can get python from a variety of places, which it seems appropriate to list without providing a recommendation: brew, MacPorts, fink, anaconda/conda

**Comment by embray** [comment]
Keep in mind these instructions are specifically about building Astropy from source.  Having instructions about this is in some ways more important for Linux than it is for OSX, since all installations that aren't through the system package manager will be installations from source.  This isn't the place to be telling OSX users not to user their system Python.  

What I'm confused about is whether or not installing Python for OSX includes the header files, and if they're installed in a way that the xcode toolchain knows about it or not.  I have no idea.

**Comment by sudheesh001** [comment]
> What I'm confused about is whether or not installing Python for OSX includes the header files,

I think its not installed by default, It seperately needs an equivalent to `python-dev` for the python development packages in OS X

**Comment by sudheesh001** [comment]
@embray @mwcraig @bsipocz Are there any other changes that you'd need me to make for this to be accepted to merge ?

**Comment by mwcraig** [comment]
Looks like the language regarding getting the headers still needs to change to not be brew-specific. Maybe something like "install python headers from a package manager like brew, macports, or fink"?

**Comment by sudheesh001** [comment]
I've made the changes :smile:

**Comment by embray** [comment]
Looks good, thanks @sudheesh001

**Pull request merged**